### PR TITLE
refactor: split local config runtime state into appropriately shielded implementations

### PR DESCRIPTION
**Pull request by nathanklick**
This pull request changes the following:

This pull request introduces a set of new interfaces and classes to manage runtime state using a "facade" pattern. It includes abstractions for collections (`FacadeArray` and `FacadeMap`), specific configuration entities (`LocalConfig`, `Deployment`, `UserIdentity`, etc.), and utility classes to encapsulate objects (`StringFacade`, `ApplicationVersions`, etc.). These changes aim to improve modularity, encapsulation, and type safety in handling runtime state.

### Collection Abstractions
* Added `FacadeArray` and `FacadeMap` interfaces to define generic collection types that wrap underlying data structures with facade objects. (`facade-array.ts`, `facade-map.ts`) [[1]](diffhunk://#diff-fcc94d0b721674ac01cc72825da0a3117c4a174e1908af1ab857b724875837f8R1-R33) [[2]](diffhunk://#diff-69c7bc31afcd3ed7ed20d1a417f2d1e5f718df6909204847d62129b9725964fcR1-R31)
* Implemented `MutableFacadeArray` and `MutableFacadeMap` classes to provide mutable implementations of the above interfaces, enabling operations like addition, removal, and iteration over facade objects. (`mutable-facade-array.ts`, `mutable-facade-map.ts`) [[1]](diffhunk://#diff-73d41831f282366402cbd9381223ac36d7b8c4b3dd3108ea7cf6fd41d2dd8c69R1-R101) [[2]](diffhunk://#diff-ceba3a8214d90ac63e74616e51ddca701f18f655c9561b76ff7fa3a9abd88587R1-R90)

### Configuration Management
* Introduced `LocalConfig` class to encapsulate the runtime state of local configurations, including deployments, user identity, and application versions. (`local-config.ts`)
* Added `Deployment` and `UserIdentity` classes to represent specific configuration entities, with getter and setter methods for encapsulated schema objects. (`deployment.ts`, `user-identity.ts`) [[1]](diffhunk://#diff-f2aa8da22403eb9d8915d172fed6ae63f56bf8dc75c8dbc191d59dddfdffc95dR1-R53) [[2]](diffhunk://#diff-9972f89dbfde26356e52d7587e30f94c397b844ae55b0982c64ea17bfbf1a942R1-R24)
* Created `LocalConfigRuntimeState` to manage the lifecycle of the local configuration, including loading, persisting, and refreshing data from a YAML backend. (`local-config-runtime-state.ts`)

### Utility Classes
* Added `ApplicationVersions` and `StringFacade` classes to encapsulate specific types of data, such as semantic versioning and string values, with additional methods for comparison and string conversion. (`application-versions.ts`, `string-facade.ts`) [[1]](diffhunk://#diff-af441beedb0f4883244adbdcb763a73a917e7dedfa4d0f8979614527327c20d9R1-R65) [[2]](diffhunk://#diff-f7853a1128b4163cb98176859f5542fc25cfe03e81368bdc12d195a483ba95dbR1-R23)
* Introduced `FacadeObjectConstructor` type to define constructors for facade objects, improving type safety in object creation. (`facade-object-constructor.ts`)

### Core Facade Interface
* Defined a `Facade` interface as a base abstraction for classes that encapsulate schema objects, ensuring a consistent API across all facade implementations. (`facade.ts`)## Description

### Related Issues

* Closes #
* Related to #

### Pull request (PR) checklist

* \[x] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[x] This PR added no TODOs or commented out code
* \[x] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[x] Any `package.json` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[x] All changes in this PR are included in the description
* \[x] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit message guidelines' below have been followed

### Testing

* \[x] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* N/A

The following was not tested:

* N/A

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any type.  NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                        |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: src/commands/deployment.ts [hunk 1]
```diff
@@ -142,13 +146,17 @@ export class DeploymentCommand extends BaseCommand {
...
-            await this.localConfig.modify(async localConfigData => {
-              localConfigData.addDeployment(deployment, namespace, realm, shard);
-            });
+            const actualDeployment: Deployment = this.localConfig.configuration.deployments.addNew();
+            actualDeployment.name = deployment;
+            actualDeployment.namespace = namespace.name;
+            actualDeployment.realm = realm;
+            actualDeployment.shard = shard;
+
+            await this.localConfig.persist();
```
Comment (instamenta) [inline src/commands/deployment.ts:159]: why are we moving away from the modify and towards an persist, i think this is a lot more error prone and a bit of an anti pattern compared to the callback

Comment (jeromy-cannon) [reply src/commands/deployment.ts:159]: The callback based approach to modification is an anti-pattern because it forces all changes to be immediately be persisted and can mask thrown errors.  

It also reduces the application/business layer's ability to determine when and if it is appropriate to persist the changes.

Being able to make a change to the runtime state without persisting the change until other components or logic has had a chance to react to the changes is an important feature. Right now to maintain the current status quo, most cases do immediately persist the changes but that will not be the case long term.

File: src/business/runtime-state/collection/mutable-facade-map.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import {type FacadeMap} from './facade-map.js';
+import {type Facade} from '../facade/facade.js';
+import {type FacadeObjectConstructor} from '../facade/facade-object-constructor.js';
+import {type ClassConstructor} from '../../utils/class-constructor.type.js';
+
+export class MutableFacadeMap<K, V extends Facade<BV>, BV> implements FacadeMap<K, V, BV> {
```
Comment (nathanklick) [inline src/business/runtime-state/collection/mutable-facade-map.ts:8]: We should see if we can eliminate the value constraint on Facade.
```suggestion
export class MutableFacadeMap<K, V> implements FacadeMap<K, V> {
```

Comment (jeromy-cannon) [reply src/business/runtime-state/collection/mutable-facade-map.ts:8]: created https://github.com/hiero-ledger/solo/issues/1975 to resolve this in to not hold this PR up.
